app/Http/Controllers/Rest/UsersController.php

```diff
@@ -0,0 +1,152 @@
+use Illuminate\Support\Facades\Validator;
+use JsonRPC\Exception\ServerErrorException;
+
+class UsersController extends AppController
+{
+    protected $email_activation_token_time = 24 * 60 * 60; // 24 hours
+    protected $email_activation_resend_limit = 15 * 60; // 15 minutes
+    protected $email_change_token_time = 60 * 60; // 1 hour
+    protected $email_new_change_resend_limit = 5 * 60; // 1 minute
+
+    /**
+     * Create a new controller instance.
+     */
+    public function __construct()
+    {
+        $this->middleware('snake_casing')->except(['emailChangeStep1']);
+    }
+
+    public function add(Request $request)
+    {
+        $this->validateRequestObject($request, 'user', User::$rules_add);
+        Validator::make($request->all(), ['uri' => 'required'])->validate();
+
+        DB::beginTransaction();
+        $user = User::register($request->input('user'));
+        Mail::to($user)->queue(
+            new UserEmailActivate(
+                Token::generate('email-activate', $this->email_activation_token_time, $user->id),
+                $request->input('uri')
+            )
+        );
+        DB::commit();
+
+        return self::json($user->toArray(), 201)
+            ->header('Location', route('app.users.read', ['user_id' => $user->id]));
+    }
+
+    public function count(Request $request)
+    {
+        throw new ServerErrorException('Not implemented yet');
+    }
+
+    public function browse(Request $request)
+    {
+        return self::json([], 501, ['message' => 'not yet implemented <3']);
+        // TODO check privileges
+        $users = User::with('AdserverWallet')->get();
+
+        return self::json($users->toArray());
+    }
+
+    public function delete(Request $request, $user_id)
+    {
+        return self::json([], 501, ['message' => 'not yet implemented <3']);
+        // TODO check privileges
+        // TODO reset email
+        // TODO process
+        $user = User::findOrFail($user_id);
+        $user->delete();
+
+        return self::json([], 204);
+    }
+
+    public function edit(Request $request)
+    {
+        if (!Auth::check() && !$request->has('user.token')) {
+            return self::json([], 401, ['message' => 'Required authenticated access or token authentication']);
+        }
+
+        DB::beginTransaction();
+        if (Auth::check()) {
+            $user = Auth::user();
+            $token_authorization = false;
+        } else {
+            if (false === $token = Token::check($request->input('user.token'), null, 'password-recovery')) {
+                DB::rollBack();
+
+                return self::json([], 422, ['message' => 'Authentication token is invalid']);
+            }
+            $user = User::findOrFail($token['user_id']);
+            $token_authorization = true;
+        }
+
+        $this->validateRequestObject($request, 'user', User::$rules);
+        $user->fill($request->input('user'));
+
+        if (!$request->has('user.password_new')) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "user.password_new" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)